python/scripts/usj2usx.py

```diff
@@ -0,0 +1,180 @@
+'''Standalone script to generate USX from input USJ'''
+
+import glob
+import argparse
+import json
+from jsonschema import validate
+import re
+import lxml.etree
+
+
+USJ_SCHEMA = None
+with open('grammar/usj.js', 'r', encoding='utf-8') as json_file:
+    USJ_SCHEMA = json.load(json_file)
+
+prev_chapter_sid = None
+prev_verse_sid = None
+
+VERSE_CARRYING_PARAS = [
+       "q", "qr", "qc", "qa", "qm", "qd", #poetry
+        "lh", "li", "lf", "lim", "litl", #lists
+        "tr", "tc", "th", "tcr", "thr", # table
+        "p", "m", "po", "pr", "cls", "pmo", "pm",
+        "pmc", "pmr", "pi", "mi", "nb", "pc", "ph", 
+        ]
+
+def find_para_node_for_verse_eid(current_usx_head):
+    if current_usx_head.tag in ["para", "cell"] and \
+        (len(current_usx_head) > 0 or current_usx_head.text is not None): # within a para
+        return current_usx_head
+    if current_usx_head.tag in ["para", "cell"] and len(current_usx_head) == 0: # start of a para
+        paras = current_usx_head.getparent().xpath('para | .//cell')[:-1]
+    else:
+        paras = current_usx_head.xpath('para | .//cell') # not in a para element
+    for i in range(len(paras)-1, -1, -1):
+        if "style" in paras[i].attrib:
+            marker = re.sub(r'\d+$', "", paras[i].attrib['style'])
+        if (paras[i].tag == "table" or marker in VERSE_CARRYING_PARAS) and \
+                (len(paras[i]) > 0 or paras[i].text is not None):
+            return paras[i]
+    raise Exception("Can't find a suitable last para:\n "+\
+        f"{lxml.etree.tostring(current_usx_head)}")
+
+
+def convert_usj(json_node, usx_head):
+    '''Traverse the JSON recursively, building the XML'''
+    global prev_verse_sid
+    global prev_chapter_sid
+    if json_node['type'] in ["chapter", "verse", "sidebar"]: # adding end nodes
+        if prev_verse_sid is not None:
+            end_node = lxml.etree.Element('verse')
+            end_node.set('eid', prev_verse_sid)
+            last_para = find_para_node_for_verse_eid(usx_head)
+            last_para.append(end_node)
+        if json_node['type'] == 'verse':
+            prev_verse_sid = json_node["sid"]
+        else:
+            prev_verse_sid = None
+        if json_node['type'] == "chapter":
+            if prev_chapter_sid is not None:
+                end_node = lxml.etree.SubElement(usx_head, 'chapter')
+                end_node.set('eid', prev_chapter_sid)    
+            prev_chapter_sid = json_node['sid']
+
```

Comment:
I think we need to append to the right place as per end verse. Also later. You made end_node, what are you going to do with it?

Reply:
I am taking start of next chapter as the place to append the previous chapter's end node. Is there a different logic. If there is I can change.